### PR TITLE
Billing - periodic operations

### PR DESCRIPTION
**Pull request by ClaudeROSSI**
Billing - Periodic Operations

- scheduled once a month (at one o'clock on the 1st day of the month) to bill invoices that are in DRAFT state.
- same job is used to attempt a payment on invoices in UNPAID state

### CODE REVIEW

**Comment by jerome-benoit** [comment]
> @jerome-benoit : I would need your advice. I introduced a new **BillingPeriodicOperationTask** task running once a month, on the 1st day at 1 AM ([crontab.guru/#0_1_1](https://crontab.guru/#0_1_1)____)
> 
> I have changed the docker/config*.json files to declare the new task (in an inactive state).
> How shall I proceed to declare it active on both our QA and PROD systems?

Yes, you have to make a PR for the configuration in ev-scp-sap-it and ev-aws repo.

File: src/migration/tasks/AddNotificationsFlagsToUsersTask.ts [hunk 1]
```diff
@@ -66,6 +66,7 @@ export default class AddNotificationsFlagsToUsersTask extends MigrationTask {
         sendSmtpError: user.notifications?.sendSmtpError ? user.notifications.sendSmtpError : false,
         sendOfflineChargingStations: user.notifications?.sendOfflineChargingStations ? user.notifications.sendOfflineChargingStations : false,
         sendBillingSynchronizationFailed: user.notifications?.sendBillingSynchronizationFailed ? user.notifications.sendBillingSynchronizationFailed : false,
+        sendBillingPeriodicOperationFailed: user.notifications?.sendBillingPeriodicOperationFailed ? user.notifications.sendBillingPeriodicOperationFailed : false,
```
Comment (jerome-benoit) [inline src/migration/tasks/AddNotificationsFlagsToUsersTask.ts:69]: You will need to also make change to the dashboard code.

File: src/integration/billing/BillingIntegration.ts [hunk 2]
```diff
@@ -177,27 +179,50 @@ export default abstract class BillingIntegration {
...
+      // Fetch OPEN invoices only
+      invoices = await BillingStorage.getInvoicesToPay(this.tenantID);
+    }
+    // Let's now finalize all invoices and attempt to get it paid
+    for (const invoice of invoices.result) {
       try {
-        await this.chargeInvoice(openInvoice);
+        // Make sure to avoid trying to charge it again too soon
+        if (moment(invoice.createdOn).isSame(moment(), 'day')) {
+          actionsDone.inSuccess++;
```
Comment (jerome-benoit) [inline src/integration/billing/BillingIntegration.ts:196]: You count not synced right now as success? Why not count is them with a specific counter like postPostponed or skipped.

File: src/integration/billing/stripe/StripeBillingIntegration.ts [hunk 3]
```diff
@@ -854,13 +862,22 @@ export default class StripeBillingIntegration extends BillingIntegration {
     };
   }
 
-  private async _getLatestDraftInvoice(customerID: string): Promise<Stripe.Invoice> {
+  private async _getLatestDraftInvoiceOfTheMonth(customerID: string): Promise<Stripe.Invoice> {
```
Comment (jerome-benoit) [inline src/integration/billing/stripe/StripeBillingIntegration.ts:865]: The underscore is not needed for private methods with TS.

File: src/integration/billing/stripe/StripeBillingIntegration.ts [hunk 4]
```diff
@@ -962,7 +988,7 @@ export default class StripeBillingIntegration extends BillingIntegration {
           source: Constants.CENTRAL_SERVER,
           action: ServerAction.BILLING_TRANSACTION,
           module: MODULE_NAME, method: 'billInvoiceItem',
-          message: `Payment attempt failed - stripe invoice: '${stripeInvoice?.id}'`,
+          message: `Payment attempt failed - stripe invoice: '${stripeInvoice?.id }'`,
```
Comment (jerome-benoit) [inline src/integration/billing/stripe/StripeBillingIntegration.ts:991]: Unneeded space

**Review by jerome-benoit — COMMENTED** [review]
Looks good to me. Just a few comments.

File: src/migration/tasks/AddNotificationsFlagsToUsersTask.ts (hunk 1 above)
Comment (ClaudeROSSI) [reply src/migration/tasks/AddNotificationsFlagsToUsersTask.ts:69]: right.
Here it is: https://github.com/sap-labs-france/ev-dashboard/issues/2412

File: src/integration/billing/BillingIntegration.ts (hunk 2 above)
Comment (ClaudeROSSI) [reply src/integration/billing/BillingIntegration.ts:196]: I agree. However the purpose of the notification is just to inform the admins when something goes wrong. Skipping invoices that have been recently created is just a normal situation.
I therefore do not think, providing additional details is worth the effort here.
Let me know what you think.

File: src/integration/billing/stripe/StripeBillingIntegration.ts (hunk 3 above)
Comment (ClaudeROSSI) [reply src/integration/billing/stripe/StripeBillingIntegration.ts:865]: Well ... I know that Serge does not like it as well. But this is helpful when looking at the code (of the calling layer). When you call something starting with "_" it just highlights the fact that you call something private (and somehow part of the core features). This helps me a lot and I would like to keep it.

File: src/integration/billing/stripe/StripeBillingIntegration.ts (hunk 4 above)
Comment (ClaudeROSSI) [reply src/integration/billing/stripe/StripeBillingIntegration.ts]: done

**Review by jerome-benoit — APPROVED** [review]

**Pull request merged**